smart-contracts/wasm-transform/src/machine.rs

```diff
@@ -571,123 +615,105 @@ impl<I: TryFromImport, R: RunnableCode> Artifact<I, R> {
-                    let cur_size = stack.size();
-                    let diff = get_u32(instructions, &mut pc);
-                    let target = get_u32(instructions, &mut pc);
-                    let top = stack.pop();
-                    if unsafe { top.short } != 0 {
-                        let top = stack.pop();
-                        stack.set_pos(cur_size - diff as usize);
-                        stack.push(top);
-                        pc = target as usize;
+                    let target = get_u32(&mut pc);
+                    let condition = get_local(constants, locals, &mut pc);
+                    if unsafe { condition.short } != 0 {
+                        pc = unsafe { instructions.as_ptr().add(target as usize) };
                     } // else do nothing
                 }
                 InternalOpcode::BrTable => {
-                    let cur_size = stack.size();
-                    let top = stack.pop();
-                    let num_labels = get_u16(instructions, &mut pc);
-                    let top: u32 = unsafe { top.short } as u32;
+                    let condition = get_local(constants, locals, &mut pc);
+                    let num_labels = get_u16(&mut pc);
+                    let top: u32 = unsafe { condition.short } as u32;
                     if top < u32::from(num_labels) {
-                        pc += (top as usize + 1) * 8; // the +1 is for the
-                                                      // default branch.
+                        pc = unsafe { pc.add((top as usize + 1) * 4) }; // the +1 is for the
+                                                                        // default branch.
                     } // else use default branch
-                    let diff = get_u32(instructions, &mut pc);
-                    let target = get_u32(instructions, &mut pc);
-                    stack.set_pos(cur_size - diff as usize);
-                    pc = target as usize;
+                    let target = get_u32(&mut pc);
+                    pc = unsafe { instructions.as_ptr().add(target as usize) };
                 }
                 InternalOpcode::BrTableCarry => {
-                    let cur_size = stack.size();
-                    let top = stack.pop();
-                    let num_labels = get_u16(instructions, &mut pc);
-                    let top: u32 = unsafe { top.short } as u32;
+                    let condition = get_local(constants, locals, &mut pc);
+                    let copy_source = get_local(constants, locals, &mut pc);
+                    let num_labels = get_u16(&mut pc);
+                    let top: u32 = unsafe { condition.short } as u32;
                     if top < u32::from(num_labels) {
-                        pc += (top as usize + 1) * 8; // the +1 is for the
-                                                      // default branch.
+                        pc = unsafe { pc.add((top as usize + 1) * 8) }; // the +1 is for the default branch.
```

Comment:
Should this not be 
```suggestion
                        pc = unsafe { pc.add((top as usize + 1) * 4) }; // the +1 is for the default branch.
```
Is there a test for this?

Reply:
On further consideration, I suspect this is correct (4 bytes for the copy target + 4 bytes for the target pc). But I find it very difficult to understand why, for `BrTable` all target labels expect no return value, but `BrTableCarry` they all expect one - though I guess it's part of the validation. Perhaps some additional documentation (also around where these op codes are generated) would be helpful.

Reply:
I added a comment to the compilation for brtable that is an attempt at explaining why we split it into two instructions.